### DISCUSSION TITLE
Pull Local Weather Data

### DISCUSSION THREAD

**Discussion by oralallen82** [jasonacox/Powerwall-Dashboard#51] [Ideas]
Utilize nearby weather station data to see how temp etc. affects production. An option to enter the Weather Underground station ID and have an option to overlay that on the production map, if that's too data intensive show highs/lows.

**Comment by jasonacox** [comment]
Hi @oralallen82 - I like this idea.  My challenge is figuring out how the best way to add this (which Weather API) as most have sign-up requirements.  I think we would want this part of the setup script.  If anyone has a good idea or suggestions, please let me know.

**Reply by jasonacox** [reply]
A new branch has been created to work on this idea.  This is also be tracking in https://github.com/jasonacox/Powerwall-Dashboard/issues/42.

Update: I have created a branch called "weather" that is my running test configuration.  Go [here](https://github.com/jasonacox/Powerwall-Dashboard/tree/weather) if you want to check it out. Keep in mind that it isn't integrated into setup.sh or upgrade.sh yet.  I'm still testing the new container.  

Here is my approach:

* I found the telegraf OpenWeatherMap integration to be glitchy (seemed to only pick up data every hour) and pivoted away from that.  It may still work for some but I wanted something more stable and granular.
* I created a new service called `weather411` that runs in a container and periodically (every 10m) reaches out to [OpenWeatherMap.org](https://openweathermap.org/) to get the latest local weather data.
* Configuration:  Will require users to sign up for a free account at [OpenWeatherMap.org ](https://openweathermap.org/)and get an API KEY.
* The API KEY, your Latitude and Longitude are entered into the [weather411.conf](https://github.com/jasonacox/Powerwall-Dashboard/blob/weather/weather/weather411.conf.sample) config file.  Ultimately this will be created by `setup.sh` or `upgrade.sh`.
* The `weather411` service stores the data in InfluxDB in the `powerwall` DB allowing us to pull this into the existing Powerwall Dashboard. 

Example overlay: 

<img width="819" alt="image" src="https://user-images.githubusercontent.com/836718/184577537-84ac5034-03b1-4a8a-8910-92c96c074914.png">

This only show the "temperature" and "clouds" (a % of cloudliness) data but there are many others available that may be interesting - I'm thinking:

* uvi (UV radiation - would be interesting to see if there was a correlation to production)
* rain (Nice to know when we get a free panel wash)
* wind (Potential cooling for higher efficiency)
* visibility (Takes into account more than just clouds but inverse metric)

Example weather data available from `weather411` that I'm graphing:

<img width="874" alt="image" src="https://user-images.githubusercontent.com/836718/184578488-b65c1410-a02c-48fa-ba0e-08c606809b1b.png">

**Reply by oralallen82** [reply]
This is really awesome. I'd definitely love to test this out. Without breaking my existing configuration.

**Comment by madkiwinz** [comment]
This is an inspired idea, I truely love it.

I would be keen on testing it and as my installation is only a few weeks old I'm less concerned about rebuilding it from scratch if required.

**Reply by jasonacox** [reply]
So far, so good!  No memory leaks:

<img width="1270" alt="image" src="https://user-images.githubusercontent.com/836718/184779057-390db218-995b-4c65-874e-d06c7b4b5431.png">

No clouds here in SoCal today so less exciting, but temp data looks good.  I'm recording and displaying the data in °C but expect several will want °F. The problem with both is that Grafana only allows for two axis (currently kW and % on the live "Energy Usage" graph).  Anyone with Grafana kung-fu skills know how to add a third (°C)?  For now, I've aligned the temperature to the % axis (-100 to 100 range):

<img width="879" alt="image" src="https://user-images.githubusercontent.com/836718/184780031-0ceb270d-864b-4869-82bb-4134baedc7c3.png">

Weather graphs from last 24hr:
<img width="1336" alt="image" src="https://user-images.githubusercontent.com/836718/184779783-18f68179-ea02-4d6a-9d68-37108f91d438.png">

**Reply by jasonacox** [reply]
@leemj001 If you want to try it out - I recommend creating a backup (e.g. `sudo tar -cvf backup.tar Powerwall-Dashboard`).  Here is what you will need to do:

```bash
# If you don't have an installation already (or want to create a new one) do this (uncomment):
# git clone https://github.com/jasonacox/Powerwall-Dashboard.git

# Go to your Powerwall-Dashboard directory
cd Powerwall-Dashboard

# Get the latest version
git pull 

# Switch to weather branch
git checkout weather
```

If you see any errors, post them here before continuing.   The next step is to configure the weather411 container.  Create a weather411.conf file and update with your specific location details:
* Enter your OpenWeatherMap API Key (APIKEY) You can get a free account and key at [OpenWeatherMap.org](https://openweathermap.org/).
* Enter your GPS Latitude (LAT) and Longitude (LON). To get your location, you can use [this tool](https://jasonacox.github.io/Powerwall-Dashboard/location.html).

Area in weather411.conf you edit:
```
[OpenWeatherMap]
# Register and get APIKEY from OpenWeatherMap.org
APIKEY = xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# Enter your location in latitude and longitude 
LAT = xxx.xxxx
LON = yyy.yyyy
```

Steps:

```bash
# Configure weather411
cp weather/weather411.conf.sample weather/weather411.conf
vi weather/weather411.conf            # Use whatever editor you prefer to edit this conf file

# Restart the stack
docker-compose -f powerwall.yml up -d
```

If all goes well, you will see it download and start the weather411 container.  Test it with `docker logs weather411` and you should see the startup header and information about your configuration.  You should also be able to visit http://localhost:8676/ (where localhost is the address of the system running your Dashboard stack) and you will see current weather data that looks like this:

<img width="304" alt="image" src="https://user-images.githubusercontent.com/836718/184781802-12163250-1374-4e64-8220-6ee2df6f33f2.png">

To add the data to your dashboard, you will need to edit the panel and two new queries (you can add more):

<img width="573" alt="image" src="https://user-images.githubusercontent.com/836718/184781965-2bc49645-0991-49ba-bb0d-2f3125e4a27d.png">

I'll upload the full json files when I have time. 

@oralallen82 - I recommend you wait for @leemj001 to test this first.  And, if you try this, please make sure you do a full backup first.  😅

**Reply by madkiwinz** [reply]
I followed the steps as you have listed and there were no issues or errors during the upgrade and docker install. 

But on the weather411 status web page http://localhost:8676/ I'm not getting any data back from the API request as shown in your screenshot example. 

I did note in the email from OpenWeather it would take a couple of hours to active the API, but its been past that.

The page only displays the following information:
Weather411 Server v0.0.5
Error: No weather data available
Last update: 1970-01-01 00:00:00
Page refresh: 2022-08-16 10:48:36.681127

When I checked the docker logs weather411 and everything looks as per your example (API key hidden) 

Weather411 Server [0.0.5]
* Configuration Loaded [/var/lib/weather/weather411.conf]
 + Weather411 - Debug: False, Activate API: True, API Port: 8676
 + OpenWeatherMap - Key: ************************, Wait: 10, Units: metric
   OpenWeatherMap - Lat: -45.87443490938819, Lon: 170.47938333566924, Timeout: 10
 + InfluxDB - Enable: True, Host: influxdb, Port: 8086, DB: powerwall, Field: weather
* Starting threads
 + fetchWeather thread
 + apiServer thread - Listening on http://localhost:8676

**Reply by mcbirse** [reply]
@jasonacox - Love this idea of logging weather data!

@leemj001 - I just tried to set this up tonight and struck the same problem.

I turned on debugging and noticed this error in the docker logs:
```
DEBUG:https://api.openweathermap.org:443 "GET /data/2.5/onecall?lat=-33.865143&lon=151.209900&units=metric&exclude=minutely,hourly,daily&appid=******************************** HTTP/1.1" 401 107
DEBUG:Error reaching OpenWeatherMap
```
When I browsed to the url the below error was returned:
```json
{"cod":401, "message": "Invalid API key. Please see http://openweathermap.org/faq#error401 for more info."}
```
I initially thought my API key was not activated yet, however the example API link I received in the e-mail was working fine.

I noticed the example link uses API call starting with `/data/2.5/weather` whereas the program is using `/data/2.5/onecall`?

Checking the [OpenWeatherMap](https://openweathermap.org/api) website, it appears they have made the `onecall` API requests a subscription service, which is charged by usage? (credit card required, up to 1,000 calls per day are free however?)

To test this, I signed up with my credit card and modified the API request url in the program as below (onecall API version changed to 3.0)
```
DEBUG:https://api.openweathermap.org:443 "GET /data/3.0/onecall?lat=-33.865143&lon=151.209900&units=metric&exclude=minutely,hourly,daily&appid=******************************** HTTP/1.1" 200 396
DEBUG:Weather data loaded
DEBUG:Writing to InfluxDB
```
Everything seems to be working fine after this and weather data is being logged. I've noted even though I have signed up with my CC, requests to the version 2.5 onecall address still do not work for me.

So, if this is a chargeable subscription service now, based on usage with 1,000 calls per day being fee, limiting the number of API calls to about once every 90 seconds should keep it free I believe.

**Reply by jasonacox** [reply]
You should only need the free subscription (that is what I have, no CC).  It took several hours before the API started working for me. I completely forgot that detail!   Here is what I signed up for:

<img width="843" alt="image" src="https://user-images.githubusercontent.com/836718/185016053-5621526d-c837-4360-a333-4e694ef6f648.png">

The API I am using came from here: https://openweathermap.org/api/one-call-3 but I didn't notice any requirement for a different subscription.  I'll check on that!

@leemj001 - Did it ever start working for you?

**Reply by madkiwinz** [reply]
Hi @jasonacox, no its still not working for me.

**Reply by mcbirse** [reply]
Hi Jason, I checked again today and the below still does not work for me. It has been >12hrs since I signed up with a new OpenWeatherMap account.

https://api.openweathermap.org/data/2.5/onecall?lat=-33.865143&lon=151.209900&units=metric&exclude=minutely,hourly,daily&appid=***********************

If I change this to below (v3.0 onecall) it is working for me, since I signed up with a subscription to the [One Call API 3.0](https://openweathermap.org/api/one-call-3) with my CC.

https://api.openweathermap.org/data/3.0/onecall?lat=-33.865143&lon=151.209900&units=metric&exclude=minutely,hourly,daily&appid=***********************

Maybe they have changed this for new users who sign up now?

**Reply by mcbirse** [reply]
Is it possible to use the `weather` API call instead of `onecall` - which I believe is completely free? Some people may be reluctant to enter CC details for the onecall API.

Below was working for me, even before signing up with a CC.

https://api.openweathermap.org/data/2.5/weather?lat=-33.865143&lon=151.209900&units=metric&exclude=minutely,hourly,daily&appid=*******

It looks like a different JSON format is returned however, so I imagine the program would require some changes for this.

**Reply by jasonacox** [reply]
This is great info.  Thanks @mcbirse and @leemj001 !  I don't know how I managed to get a free account that works with the /2.5/onecall - https://api.openweathermap.org/data/2.5/onecall?lat=-33.865143&lon=151.209900&units=metric&exclude=minutely,hourly,daily&appid=xxx

<img width="662" alt="image" src="https://user-images.githubusercontent.com/836718/185019194-485484fe-c4d6-4065-82d8-3b39ac20f33b.png">

But I can't use the 3.0 version - I get: {"cod":401, "message": "Invalid API key. Please see http://openweathermap.org/faq#error401 for more info."}

I'll pivot to /api.openweathermap.org/data/2.5/weather and update the code to make use of that.  I think it has most of the relevant data points we would want.  

@leemj001 can you test this url with your API key just to confirm that you don't need a CC account?  

https://api.openweathermap.org/data/2.5/weather?lat=-33.865143&lon=151.209900&units=metric&exclude=minutely,hourly,daily&appid=

**Reply by mcbirse** [reply]
> I'll pivot to /api.openweathermap.org/data/2.5/weather and update the code to make use of that. I think it has most of the relevant data points we would want.

Sounds good.

I just checked, and it appears the "weather" call is missing these two data points compared to onecall: "dew_point" and "uvi"

Also, I think with the "weather" call the `&exclude=minutely,hourly,daily` is not required and makes no difference.

**Reply by jasonacox** [reply]
Thanks @mcbirse - good points!  It also looks like the 'alerts' are not going to show up in weather but I think that is fine as well.  Based on the https://openweathermap.org/current definitions, I noticed there is also data on snowfall. I added that as well thought I don't think I will see much data on that in SoCal. 😁  Still, I believe it could be useful information for solar.

I made the updates tonight and uploaded the new version (0.1.0) of weather411.  Quick instructions to pull down the latest:

```bash
# Remove old version
docker stop weather411
docker rm weather411
docker images | grep weather411 | awk '{print $3}' | xargs docker rmi -f

# Restart the stack
docker-compose -f powerwall.yml up -d
```

Check this for current weather data: http://localhost:8676/ and `docker logs weather411`.

**Reply by madkiwinz** [reply]
I've updated to version 0.1.0 and data is displayed in the test page http://localhost:8676/

I haven't updated the panel yet, work is being demanding this week :)

**Reply by mcbirse** [reply]
I have updated to v0.1.0 as well and all is working well! Awesome work Jason. :thumbsup:

I submitted a pull request to remove `/dew_point` from the available API url's of Weather411 as I noticed it was still available but will return no data. Hope that pull request is okay... still learning GitHub and was not sure how to submit a PR to a branch?

e.g. http://localhost:8676/dew_point returns an error
![image](https://user-images.githubusercontent.com/108725631/185153067-7278a608-80ac-4794-8a4a-991d09dfd615.png)

**Reply by jasonacox** [reply]
Great catch @mcbirse .  Thank you!  I'm going to move on to updating the `setup.sh` and `update.sh` scripts.  Thanks for all your help!

**Reply by jasonacox** [reply]
I completed a first pass at the updated scripts.  I added `weather.sh` that will be used by the `setup.sh` and `update.sh` scripts, but can also work independently.  

I posted this in issue #42 to welcome more beta testers.  

## BETA Testers 

...or should I say, "Storm chasers"? 😄 

If you want to try the weather data feature, here is how you can check it out.  

**Highly recommended**: back up your data first.

```bash
# from the Powerwall-Dashboard folder

# First pull down the latest code from Github
git pull

# Second, switch to the weather branch (the BETA code)
git checkout weather

# Third, run the weather setup script and follow instructions
./weather.sh
```

## OpenWeatherMap Note

After you sign up for the free OpenWeatherMap.org account, make sure you check your email and verify.  Also, the API Key you get may not be active right away.  It could take up to 2 hours but in my recent test it was working within 10 minutes.  Please let me know your experience! 

## Dashboard Update

You need to install the newer `dashboard.json` that includes the weather data for the "live monitoring" graphs. In Grafana:

* From `Dashboard\Manage` (or `Dashboard\Browse`), select `Import`
* Upload one the `dashboard.json` in your Powerwall-Dashboard folder.
* If you want a weather dashboard, you can upload the test dashboard `dashboard-weather.json` in the `weather/` folder.

Hopefully you will soon see something like this:

<img width="791" alt="image" src="https://user-images.githubusercontent.com/836718/185774057-acdf5dd1-a85e-4c97-bb7b-fe8f294404f8.png">

<img width="1128" alt="image" src="https://user-images.githubusercontent.com/836718/185774521-676a3f51-1ad7-4d81-af7d-0f26d8d4668c.png">

## Issues

* The temperature data on the Live monitoring graph shows up as (%) instead of °C or °F.  From what I can tell, Grafana only allows two axis.  We have "kW" and "%".  I put temperature on the "%" axis but it is confusing. If anyone has a better idea, please let me know.
* The graphs are basically set for °C but during the weather setup (weather.sh) you can choose Imperial units (°F) which is what will show up on the live monitoring instead of °C.  However, for the weather dashboard, the values will be wrong as the axis is set for °C.  Edit the panel and change the units from °C to °F. 
* The weather411 service is storing all of the data from OpenWeatherMap into InfluxDB so there are many other data points you can play with. The weather411 graphs above show some of it.  

## How's the weather out there? 😁 

Before I release this as `v2.5.0` officially, I am running test and looking for feedback from the community.  Please post any issues, questions and suggestions here.

**Comment by oralallen82** [comment]
Good progress so far. Unfortunately I'm as much help on Grafana as a blind
pilot 😁.

**Reply by jasonacox** [reply]
No worries! Posting here for anyone else who might know.  It works as-is.  I'm just a bit OCD about units. 😉

**Comment by youzer-name** [comment]
I added weather411 to my Powerwall Dashboard and am seeing errors in the stats:

```
{"weather411": "0.1.2", "gets": 10, "errors": 5, "timeout": 0, "uri": {"/stats": 5}, "ts": 1661781734, "start": 1661781707, "clear": 1661781707, "influxdb": 1, "influxdberrors": 0, "mem": 15056}
```

Every time I refresh, the 'gets' goes up by 2 and the 'errors' goes up by 1. Should I be concerned?

When I first start the weather411 container I get 3 attempts to connect to influxdb and then an error writing to influxdb. Is this because the database container isn't up and ready before the first attempts to write to it? If I just stop and start the weather411 container, I don't get the inital write error (but I still get the errors in the stats output).


My installation is customized and I don't use the install scripts. I have to merge all the changes into my existing setup, so I may have messed something up when adding the new features.

**Reply by jasonacox** [reply]
Hi @youzer-name!  Are you seeing the weather data in Grafana?

The errors you see show up in the stats are due to the browser calling for favicon.ico which is not a valid API call for weather411.  I should probably just add that URL as valid and send back a blank file. 

Yes, weather411 needs influxdb to be up and stable before it starts. It is the same as telegraf.  The error is non-fatal and weather411 will keep trying.  If you want to add the delay yourself, you can use a bash script like this (from `setup.sh`) to wait for influxdb to be up before starting weather411:

```bash
# Service Running Helper Function
running() {
    local url=${1:-http://localhost:80}
    local code=${2:-200}
    local status=$(curl --head --location --connect-timeout 5 --write-out %{http_code} --silent --output /dev/null ${url})
    [[ $status == ${code} ]]
}

# Wait for influxdb
echo "Waiting for InfluxDB to start..."
until running http://localhost:8086/ping 204 2>/dev/null; do
    printf '.'
    sleep 5
done
echo " up!"

# Start weather411 and telegraf etc.
```

**Reply by youzer-name** [reply]
Thanks @jasonacox - Yes I'm seeing the data in Grafana now. I had previously been displaying OpenWeather data from another app's local database in Grafana, but it's nice to have this all self-contained now.

**Comment by oralallen82** [comment]
![image](https://user-images.githubusercontent.com/33994774/188911319-2827ac73-b859-46e4-b40c-10760d5f7e8b.png)
Should that be real feel?

**Reply by mcbirse** [reply]
So the reported "feels like" temperature is higher than actual, correct?

Do you have high humidity at the moment? I think that is perfectly valid and possible if you also have very high humidity. Would be interesting to know what the humidity percentage shows in bottom weather panel?

**Comment by oralallen82** [comment]
The reason I asked is that there isn't an indication for this on
the dashboard. I have temperature max and solar, nothing for "Real feel" I
am aware that humidity plays a factor in the real feel. Growing up in the
Caribbean I can attest to that.

**Reply by mcbirse** [reply]
You mean in the list of values when you hover the "feels like" value is not shown, and it is not shown in the "max" bottom values?

So, you could edit the panel to add those - I had already done that in mine so I could see the feels like temp:
![image](https://user-images.githubusercontent.com/108725631/189006179-9ab4af0a-521d-438a-b998-de24e3e6eecb.png)

Or maybe suggest this as an improvement to be added to the default panel JSON if everyone thinks this is an improvement? (I do)

If you want to do that yourself though, edit the Energy Usage panel:
![image](https://user-images.githubusercontent.com/108725631/189006676-1cb5e4b1-26f0-4bb7-a7d2-6a88817b0300.png)
Select Visualization:
![image](https://user-images.githubusercontent.com/108725631/189006701-2fa74910-82a6-44f7-b296-ef875b9cd7e6.png)
Then find the "Feels Like" series and remove the "Legend: false" and "Hide in tooltip: true" by clicking the X next to each:
![image](https://user-images.githubusercontent.com/108725631/189006769-17e6a7a7-4e9a-4ff7-894d-7f8a976c1282.png)

**Reply by oralallen82** [reply]
I'll follow your lead and edit myself. It would be a good enhance for others who may not be interested in manually editing. I know @jasonacox  has done a great deal of work from my request to beta then live.

**Reply by mcbirse** [reply]
Yes, I think it would be a good enhancement personally. There is an argument for "information overload" perhaps - but I get the feeling most people using the dashboard like to see more data, not less. 😄 

I also changed the order of mine to Clouds, Temp, Feels Like, as it made more sense - but that was done by editing the JSON directly:
![image](https://user-images.githubusercontent.com/108725631/189008316-786f3629-e73c-4682-ae52-8742972807f2.png)

**Reply by oralallen82** [reply]
That does make sense how you have it ordered, I may take a stab at it and see if I can pull off editing the .JSON
